src/server/game/Entities/Object/Object.cpp

```diff
@@ -2856,7 +2856,10 @@ SpellCastResult WorldObject::CastSpell(CastSpellTargetArg const& targets, uint32
 
     spell->m_customArg = args.CustomArg;
 
-    return spell->prepare(*targets.Targets, args.TriggeringAura);
+    if (args.Delay > 0ms)
+        this->m_Events.AddEventAtOffset([spell, targets, &args]() { return spell->prepare(*targets.Targets, args.TriggeringAura); }, args.Delay);
```

Comment:
Could you test this with a delay of 2 secs?

Reply:
This is not good, you are passing args to delayed lambda by reference (its a local variable in the function calling CastSpell), it will point to random garbage when it executes

Reply:
Yeah, @mdX7 already told me it's not a good idea to pass args on this. I'd use the m_ variable directly, the issue is TriggeringAura changes depending on the function precious to this one 🫥

Reply:
It gets worse than that, TriggeringAura might become invalid too

Reply:
In that case, how is it possible to pass it without ever being invalid?

Reply:
Invert the logic, delay the call to CastSpell

Reply:
Yeah, this would go ahead of my knowledge to handle properly if it went into that direction. What about a new method for WorldObject::CastSpellWithDelay which calls for CastSpell directly  through m_Event and make sure we always pass an ObjectGuid if target is an unit to be used in the lambda?